## Applying Boosting technique and comparing against DecisionTreeClassifier

In [1]:
#Import necessary libraries
import pandas as pd
import numpy as np
import pickle
from sklearn import datasets
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score
import warnings
warnings.filterwarnings('ignore')

In [2]:
# reading the features and the labels
data= pd.read_csv('pima-indians-diabetes.csv')

In [3]:
data.head()

Number of times pregnant  Plasma glucose concentration  \
0                         6                           148   
1                         1                            85   
2                         8                           183   
3                         1                            89   
4                         0                           137   

   Diastolic blood pressure (mm Hg)  Triceps skinfold thickness (mm)  \
0                                72                               35   
1                                66                               29   
2                                64                                0   
3                                66                               23   
4                                40                               35   

   2-Hour serum insulin (mu U/ml)  \
0                               0   
1                               0   
2                               0   
3                              94   
4                             168   

   Body mass index (weight in kg/(height in m)^2)  Diabetes pedigree function  \
0                                            33.6                       0.627   
1                                            26.6                       0.351   
2                                            23.3                       0.672   
3                                            28.1                       0.167   
4                                            43.1                       2.288   

   Age  Is Diabetic  
0   50            1  
1   31            0  
2   32            1  
3   21            0  
4   33            1

In [4]:
data.shape

(768, 9)

In [5]:
data.columns

Index(['Number of times pregnant', 'Plasma glucose concentration',
       'Diastolic blood pressure (mm Hg)', 'Triceps skinfold thickness (mm)',
       '2-Hour serum insulin (mu U/ml)',
       'Body mass index (weight in kg/(height in m)^2)',
       'Diabetes pedigree function', 'Age', 'Is Diabetic'],
      dtype='object')

In [6]:
cols = ['Plasma glucose concentration',
       'Diastolic blood pressure (mm Hg)', 'Triceps skinfold thickness (mm)',
       '2-Hour serum insulin (mu U/ml)',
       'Body mass index (weight in kg/(height in m)^2)',
       'Diabetes pedigree function', 'Age']

In [7]:
# Replacing zeroes with nan
for col in cols:
    data[col]=data[col].replace(0, np.nan)

In [8]:
# checking for missing values
data.isna().sum()

Number of times pregnant                            0
Plasma glucose concentration                        5
Diastolic blood pressure (mm Hg)                   35
Triceps skinfold thickness (mm)                   227
2-Hour serum insulin (mu U/ml)                    374
Body mass index (weight in kg/(height in m)^2)     11
Diabetes pedigree function                          0
Age                                                 0
Is Diabetic                                         0
dtype: int64

In [9]:
# imputing the missing values
data['Plasma glucose concentration']=data['Plasma glucose concentration'].fillna(data['Plasma glucose concentration'].mode()[0])
data['Diastolic blood pressure (mm Hg)']=data['Diastolic blood pressure (mm Hg)'].fillna(data['Diastolic blood pressure (mm Hg)'].mode()[0])
data['Triceps skinfold thickness (mm)']=data['Triceps skinfold thickness (mm)'].fillna(data['Triceps skinfold thickness (mm)'].mean())
data['2-Hour serum insulin (mu U/ml)']=data['2-Hour serum insulin (mu U/ml)'].fillna(data['2-Hour serum insulin (mu U/ml)'].mean())
data['Body mass index (weight in kg/(height in m)^2)']=data['Body mass index (weight in kg/(height in m)^2)'].fillna(data['Body mass index (weight in kg/(height in m)^2)'].mean())


In [10]:
# checking for missing values after imputation
data.isna().sum()

Number of times pregnant                          0
Plasma glucose concentration                      0
Diastolic blood pressure (mm Hg)                  0
Triceps skinfold thickness (mm)                   0
2-Hour serum insulin (mu U/ml)                    0
Body mass index (weight in kg/(height in m)^2)    0
Diabetes pedigree function                        0
Age                                               0
Is Diabetic                                       0
dtype: int64

In [11]:
#Separating the feature and the Label columns 
x=data.drop(labels='Is Diabetic', axis=1)
y= data['Is Diabetic']

In [12]:
# as the datapoints differ a lot in magnitude, we'll scale them using StandardScaler

scaler = StandardScaler()
scaled_data = scaler.fit_transform(x)

In [13]:
from sklearn.model_selection import train_test_split
train_x,test_x,train_y,test_y=train_test_split(scaled_data,y,test_size=0.3,random_state=42)

In [14]:
# fit DecisionTreeClassifier to training data
model = DecisionTreeClassifier()
model.fit(train_x, train_y)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [15]:
# cheking training accuracy
y_pred = model.predict(train_x)
predictions = [round(value) for value in y_pred]
accuracy = accuracy_score(train_y,predictions)
accuracy

1.0

In [16]:
# cheking initial test accuracy
y_pred = model.predict(test_x)
predictions = [round(value) for value in y_pred]
accuracy = accuracy_score(test_y,predictions)
accuracy

0.6666666666666666

## Using AdaBoostClassifier

In [17]:
# Instantiate Adaboost with weak classifier (DecisionTreeClassifier is our weak classifier i.e model)
ada = AdaBoostClassifier(base_estimator=model)

In [18]:
ada.fit(train_x,train_y)
y_pred= ada.predict(test_x)

In [20]:
print ('accuracy = > ', accuracy_score(test_y,y_pred))

accuracy = >  0.696969696969697


## Hyperparameter tuning for Adaboost using GridSearchCV

In [72]:
params = {'n_estimators':[20,25,35], 'learning_rate':[0.2,0.25,0.3]}

In [73]:
grd_ada = GridSearchCV(AdaBoostClassifier(random_state=42),param_grid=params)

In [74]:
grd_ada.fit(train_x,train_y)

GridSearchCV(cv=None, error_score=nan,
             estimator=AdaBoostClassifier(algorithm='SAMME.R',
                                          base_estimator=None,
                                          learning_rate=1.0, n_estimators=50,
                                          random_state=42),
             iid='deprecated', n_jobs=None,
             param_grid={'learning_rate': [0.2, 0.25, 0.3],
                         'n_estimators': [20, 25, 35]},
             pre_dispatch='2*n_jobs', refit=True, return_train_score=False,
             scoring=None, verbose=0)

In [75]:
grd_ada.best_estimator_

AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None, learning_rate=0.3,
                   n_estimators=35, random_state=42)

In [76]:
ada = AdaBoostClassifier(base_estimator=model,learning_rate=0.3,n_estimators=35)

In [77]:
ada.fit(train_x,train_y)
y_pred= ada.predict(test_x)

In [78]:
print ('accuracy post tuning = > ', accuracy_score(test_y,y_pred))

accuracy post tuning = >  0.6796536796536796


With above params we are able to reach around 68%. We can still finetune the parameters to get good accuracy. 
- One can tune DecisionTreeClassifier hyperparams and get the new model and then use that new model as base estimator in AdaBoostClassifier to get better accuracy